src/main/java/com/faforever/client/fx/PlayerContextMenuController.java

```diff
@@ -0,0 +1,249 @@
+
+@Slf4j
+@Scope(ConfigurableBeanFactory.SCOPE_PROTOTYPE)
+@Component
+@RequiredArgsConstructor
+public abstract class PlayerContextMenuController implements Controller<ContextMenu> {
+
+  protected final EventBus eventBus;
+  private final JoinGameHelper joinGameHelper;
+  protected final ModeratorService moderatorService;
+  private final NotificationService notificationService;
+  private final PlayerService playerService;
+  private final ReplayService replayService;
+  protected final UiService uiService;
+
+  public ContextMenu playerContextMenuRoot;
+  public MenuItem showUserInfo;
+  public MenuItem sendPrivateMessageItem;
+  public MenuItem copyUsernameItem;
+  public CustomMenuItem colorPickerMenuItem;
+  public ColorPicker colorPicker;
+  public Button removeCustomColorButton;
+  public SeparatorMenuItem socialSeparator;
+  public MenuItem inviteItem;
+  public MenuItem addFriendItem;
+  public MenuItem removeFriendItem;
+  public MenuItem addFoeItem;
+  public MenuItem removeFoeItem;
+  public MenuItem reportItem;
+  public SeparatorMenuItem gameSeparator;
+  public MenuItem joinGameItem;
+  public MenuItem watchGameItem;
+  public MenuItem viewReplaysItem;
+  public SeparatorMenuItem moderatorActionSeparator;
+  public MenuItem kickGameItem;
+  public MenuItem kickLobbyItem;
+  public MenuItem broadcastMessage;
+  public CustomMenuItem avatarPickerMenuItem;
+  public ComboBox<AvatarBean> avatarComboBox;
+
+  private PlayerBean player;
+  private InvalidationListener playerPropertyInvalidationListener;
+
+  @Override
+  public void initialize() {
+    showUserInfo.setOnAction(event -> onShowUserInfoSelected());
+    copyUsernameItem.setOnAction(event -> onCopyUsernameSelected());
+    addFriendItem.setOnAction(event -> onAddFriendSelected());
+    removeFriendItem.setOnAction(event -> onRemoveFriendSelected());
+    addFoeItem.setOnAction(event -> onAddFoeSelected());
+    removeFoeItem.setOnAction(event -> onRemoveFoeSelected());
+    reportItem.setOnAction(event -> onReport());
+    joinGameItem.setOnAction(event -> onJoinGameSelected());
+    watchGameItem.setOnAction(event -> onWatchGameSelected());
+    viewReplaysItem.setOnAction(event -> onViewReplaysSelected());
+    kickGameItem.setOnAction(event -> onKickGame());
+    kickLobbyItem.setOnAction(event -> onKickLobby());
+
+    initializeListeners();
+
+    socialSeparator.visibleProperty().bind(addFriendItem.visibleProperty().or(
+        removeFriendItem.visibleProperty().or(
+            addFoeItem.visibleProperty().or(
+                removeFoeItem.visibleProperty()))));
+    gameSeparator.visibleProperty().bind(joinGameItem.visibleProperty().or(
+        watchGameItem.visibleProperty().or(
+            viewReplaysItem.visibleProperty())));
+    moderatorActionSeparator.visibleProperty().bind(kickGameItem.visibleProperty().or(
+        kickLobbyItem.visibleProperty().or(
+            broadcastMessage.visibleProperty())));
```

Comment:
Would be good to have the closing parentheses match the level of the open parentheses. Would make it easier to see